[Buildkite] Test Android Staging on `ami-0953ca1d949c23a58` -> `ami-0f976e0b4de96aa28`

ParaskP7 opened the pull request:
Related: [a8c-ci-toolkit-buildkite-plugin#142](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/142)

AMI Name: `android-build-image-6.12.0v1.5-rc-1`

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR changes the Buildkite agent from `android` to `android-staging`. This change is not meant to be merged, but rather used to verify that the new AMI `ami-0953ca1d949c23a58` works as expected.

---

### Before the Change

---

- [Saving Gradle dependency cache...](https://buildkite.com/automattic/woocommerce-android/builds/26681#01947bde-9c0b-4a45-ba3f-fab98d0702e9/610-611)
  - Duration: [0 minutes and 43 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26681#01947bde-9c0b-4a45-ba3f-fab98d0702e9/610-621)
  - Cache Size: [628.98 MB](https://buildkite.com/automattic/woocommerce-android/builds/26681#01947bde-9c0b-4a45-ba3f-fab98d0702e9/610-622)
- Build Time: [20m 28s](https://buildkite.com/automattic/woocommerce-android/builds/26681#01947bde-9c0b-4a45-ba3f-fab98d0702e9)

-----

- [Restoring Gradle dependency cache...](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/185-186)
  - Duration: [0 minutes and 14 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/185-194)
  - Cache Size: [628.98 MB](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/185-195)
- [Build](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474) (`Lint`): [Build Scan](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/199-258) -> [Network Activity*](https://scans.gradle.com/s/axqj2agmid3dm/performance/network-activity)
  - Files Downloaded: `12`
  - Network Requests: `37`
  - Time Spent: `4s`

(*) Although, note that these numbers might not be very representative because `trunk` is using a dependency cache that is already outdated (5 days has passed since it was last updated, every Sunday).

---

### After the Change

---

- [Saving Gradle dependency cache...](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09/17548-17549)
  - Duration: [1 minutes and 12 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09/17548-17559)
  - Cache Size: [950.32 MB](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09/17548-17560)
- Build Time: [7m 14s](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09)

-----

- [Restoring Gradle dependency cache...](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/187-188)
  - Duration: [0 minutes and 19 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/187-196)
  - Cache Size: [950.32 MB](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/187-197)
- [Build](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7) (`Lint`): [Build Scan](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/201-9143) -> [Network Activity](https://scans.gradle.com/s/hwykrfpxihn2a/performance/network-activity)
  - Files Downloaded: `0`
  - Network Requests: `0`
  - Time Spent: `0`

---

### Conclusions

As expected, using `./gradlew downloadDependencies androidDependencies` is:
1. Simpler, much less configuration and reasoning about which tasks to use.
2. Scalable, no need to add any more tasks to the `save-cache.sh` script as the project evolves.
3. Producing a bigger cache,
    - Before: [628.98 MB](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/185-195) 
    - After: [950.32 MB](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/187-197)
    - Result: `+ 321.34 MB` more data downloaded, which means even less network requests on CI.
4. Slower, both in saving and restoring the cache, which is the only drawable:
    - Save:
        - Before: [0 minutes and 43 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26681#01947bde-9c0b-4a45-ba3f-fab98d0702e9/610-621)
        - After: [1 minutes and 12 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09/17548-17559)
        - Result: `+ 29 seconds` to save the cache.
    - Restore:
        - Before: [0 minutes and 14 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26799#019497c6-3be1-4c32-a8b9-253181ac7474/185-194)
        - After: [0 minutes and 19 seconds](https://buildkite.com/automattic/woocommerce-android/builds/26805#01949883-7a9b-4a0a-bfe3-bec2fb1eb2c7/187-196)
        - Result: `+ 5 seconds` to restore the cache.

ParaskP7 committed:
CI: Use android-staging agent to test ami-0953ca1d949c23a58

AMI Name: android-build-image-6.12.0v1.5-rc-1

ParaskP7 committed:
CI: Update a8c-ci-toolkit to pr-sha commit (test download dependencies)

FYI: This change is done for testing purposes only, and in order to
verify that the newly added 'init.gradle.kts' script is working as
expected.

A8C CI Toolkit PR: [Dependency Cache] Test Custom Init Gradle Task
to Download Dependencies #142
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/142

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
CI: Replace all individual assemble tasks with download dependencies

This custom 'downloadDependencies' task is part of the Android AMI,
that is utilized by the 'android' agents, and it is becoming available
to all project via the 'init.gradle.kts' script that is included in it.

ParaskP7 commented:
👋 @wzieba , I just added you as a review on this draft PR, to be closed when we get ready with the new Android AMI, but please, take a look at the description (results). 🙏

TL;DR: Everything is working as expected, a total win, I am happy with the result, with the only drawback that it now takes more time to save and restore the cache.
- During saving, this is actually irrelevant as the total build time dropped anyway.
- But, during restoring, you add an additional `5s` to `10s`, which is indeed something we expected, but worth noticing/thinking about anyway. 👀

wzieba commented on .buildkite/commands/save-cache.sh:
```diff
@@ -2,40 +2,8 @@
-echo "--- 🧪 Download Android Test Dependencies [Assemble Android Tests]"
-./gradlew assembleJalapenoDebugAndroidTest
+echo "--- 📦 Download Dependencies"
+./gradlew downloadDependencies androidDependencies
```
WDYT about adding the task dependency to `downlaodDependencies` so we don't have to remember about running `androidDependencies` every time, and have this as one-task?

I mean something like this in the initialization script

```groovy
tasks.named('downloadDependencies').configure {
    finalizedBy 'androidDependencies'
}
```

and then here

```suggestion
./gradlew downloadDependencies
```

ParaskP7 replied:
Oh, yea, this is a great idea @wzieba , thanks for suggestion! 💯 

Btw, I thought about the pros/cons of doing something like that, for example:

- Cons: Being implicit, not flexible. 🤔
- Pros: When and if need be, we will only update the Android AMI to add/remove extra tasks from this `downloadDependencies`, without the need to open multiple PRs on every client/library project. 🥇 

And I think this single `pros` make this change worthy it. 👍 

FYI: I'll jump right on it, create another staging Android AMI to quickly test this change on this and the [JP/WPAndroid#21609](https://github.com/wordpress-mobile/WordPress-Android/pull/21609) PR, just to be safe, and before proceeding with making all the related PRs ready for review, starting with [buildkite-ci#585](https://github.com/woocommerce/woocommerce-android/pull/13387#discussion_r1931721646).

ParaskP7 replied:
This above is now 2f4941b531b68e0b86ae0da77fd6eab9eefbfdd8 ✅ @wzieba and everything is looking as expected:

- Before: [950.32 MB](https://buildkite.com/automattic/woocommerce-android/builds/26803#01949877-fadf-444d-b544-e5a13e487a09/17548-17560) cache size -> [4009](https://scans.gradle.com/s/ffvlzfhqdddko/performance/network-activity) network requests
- After: [950.29 MB](https://buildkite.com/automattic/woocommerce-android/builds/26851#0194ac81-0604-4207-8292-90b8d2eb0da8/17554-17566) cache size -> [4009](https://scans.gradle.com/s/zr2pluvvzemme/performance/network-activity) network requests

wzieba replied:
Thanks! 🙌

wzieba reviewed (approved):
> But, during restoring, you add an additional 5s to 10s, which is indeed something we expected, but worth noticing/thinking about anyway. 👀

Yeah, thanks for highlighting this! It's unfortunate and might even change in which jobs we use the dependency cache, but so low cost of maintaining the "save-cache" script across projects is I think worth these 5-10s.

ParaskP7 commented:
> Yeah, thanks for highlighting this! It's unfortunate and might even change in which jobs we use the dependency cache, but so low cost of maintaining the "save-cache" script across projects is I think worth these 5-10s.

Btw @wzieba , forgot to say that I agree with what you say here, and worse case we could revert back to using the `assemble` tasks for saving the cache, or just the `assembleDebugAndroidTest`, and switching between solutions as we see fit. 👍

ParaskP7 committed:
CI: Remove unncessary android dependencies from download dependencies

A new Android AMI got created that now finalizes 'downloadDependencies'
with the 'androidDependencies' task.
